### PR TITLE
[feature](pipelineX) Use dependency instead of block queue in the runtime filter

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

File: be/src/exprs/runtime_filter.h [hunk 1]
```diff
@@ -384,6 +388,17 @@ class IRuntimeFilter {
         }
     }
 
+    int32_t wait_time_ms() {
+        auto runtime_filter_wait_time_ms = _state == nullptr
```
Comment (Gabriel39) [inline be/src/exprs/runtime_filter.h:392]: Always set `_query_ctx` and delete all of these judgement

File: be/src/vec/exec/runtime_filter_consumer.cpp [hunk 2]
```diff
@@ -26,7 +26,9 @@ RuntimeFilterConsumer::RuntimeFilterConsumer(const int32_t filter_id,
         : _filter_id(filter_id),
           _runtime_filter_descs(runtime_filters),
           _row_descriptor_ref(row_descriptor),
-          _conjuncts_ref(conjuncts) {}
+          _conjuncts_ref(conjuncts) {
+    _blocked_by_rf = std::make_shared<std::atomic_bool>(false);
```
Comment (Gabriel39) [inline be/src/vec/exec/runtime_filter_consumer.cpp:30]: _blocked_by_rf(false)

File: be/src/pipeline/pipeline_x/dependency.h [hunk 3]
```diff
@@ -194,30 +199,67 @@ class FinishDependency final : public Dependency {
...
+    const int32_t _wait_time_ms;
+    IRuntimeFilter* _runtime_filter;
+};
+class RuntimeFilterDependency final : public Dependency {
+public:
+    RuntimeFilterDependency(int id, int node_id, std::string name)
+            : Dependency(id, name), _node_id(node_id) {}
+
+    RuntimeFilterDependency* filter_blocked_by() {
+        if (!_blocked_by_rf) {
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/dependency.h:241]: if (!_blocked_by_rf || _filters == 0) {

File: be/src/pipeline/pipeline_x/operator.cpp [hunk 4]
```diff
@@ -315,7 +315,7 @@ PipelineXLocalStateBase::PipelineXLocalStateBase(RuntimeState* state, OperatorXB
           _state(state),
           _finish_dependency(new FinishDependency(parent->operator_id(), parent->node_id(),
                                                   parent->get_name() + "_FINISH_DEPENDENCY")) {
-    _filter_dependency = std::make_unique<FilterDependency>(
+    _filter_dependency = std::make_shared<RuntimeFilterDependency>(
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/operator.cpp:318]: `_filter_dependency` is only be hold in PipelineXLocalStateBase. So is it enough to use unique pointer?

File: be/src/pipeline/pipeline_x/dependency.h [hunk 5]
```diff
@@ -194,30 +199,67 @@ class FinishDependency final : public Dependency {
...
+        /// TODO:
+    }
+    void set_blocked_by_rf(std::shared_ptr<std::atomic_bool> blocked_by_rf) {
+        _blocked_by_rf = blocked_by_rf;
     }
 
 protected:
-    std::function<bool()> _runtime_filters_are_ready_or_timeout;
     const int _node_id;
+    std::atomic_int _filters;
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/dependency.h:261]: We use both `_filters` and `_blocked_by_rf` to find out whether this dependency is ready. But I think it is enough to use either one

File: be/src/pipeline/pipeline_x/dependency.cpp [hunk 6]
```diff
@@ -326,4 +330,116 @@ Status HashJoinDependency::extract_join_column(vectorized::Block& block,
...
+    }
+    _call_timeout = true;
+    if (_parent) {
+        _parent->sub_filters();
+    }
+}
+
+void RuntimeFilterTimer::call_ready() {
+    std::unique_lock<std::mutex> lc(_lock);
+    /// TODO: check _runtime_filter->is_ready_or_timeout();
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/dependency.cpp:352]: Do not to call `is_ready_or_timeout` because we already use dependency to replace it

**Comment by Mryange** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**